File: LeetCode/Java/Medium/DailyTemperatures.java
/**
 * 739. Daily Temperatures
 * Difficulty: Medium
 * Description: Given an array of integers temperatures represents the daily temperatures, return an array
 * answer such that answer[i] is the number of days you have to wait after the ith day to get a warmer temperature.
 * If there is no future day for which this is possible, keep answer[i] == 0 instead.
 */
public class DailyTemperatures {
    public int[] dailyTemperatures(int[] temperatures) {
        // Monotonic decreasing stack for finding next greater element
        int[] res = new int[temperatures.length];
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < temperatures.length; i++) {
            // Stack not empty and the top of the stack is smaller than current temp.
            while (!stack.isEmpty() && (temperatures[stack.peek()] < temperatures[i])) {
                int topIndex = stack.pop();
                res[topIndex] = i - topIndex;
            }
            stack.push(i);
        }

        return res;
    }
}

File: LeetCode/Python/Medium/daily_temperatures.py
#
#   739. Daily Temperatures
#   Difficulty: Medium
#   Description: Given an array of integers temperatures represents the daily temperatures, return an array answer such that answer[i] is the number #   of days you have to wait after the ith day to get a warmer temperature. If there is no future day for which this is possible, keep answer[i] == 0 #   instead.

#   Input: temperatures = [73,74,75,71,69,72,76,73]
#   Output: [1, 1, 4, 2, 1, 1, 0, 0]

#   Input: temperatures = [30,40,50,60]
#   Output: [1, 1, 1, 0]

from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = []

        for idx, temp in enumerate(temperatures):
            while stack and temp > stack[-1][0]:
                stack_temp, stack_idx = stack.pop()
                res[stack_idx] = idx - stack_idx
            stack.append((temp, idx))

        return res
